Dependency Upgrades

AlanCoding opened the pull request:
More notes in

https://gist.github.com/AlanCoding/9442a512ab6977940bc7b5b346d4f70b

One of the upgrades I'm more unsure about is Azure. There's probably more history there than I'm aware of.

replaces #1198

Commit:
Constrain requests

AlanCoding committed:
Dependency Updates

Upgrades of minor dependency upgrades
Inventory scripts were upgraded in separate commit

Major exclusions from this update
- celery was already downgraded for other reasons
- Django / DRF major update already done, minor bumps here
- asgi-amqp has fixes coming independently, not touched
- TACACS plus added features not needed

Removals of note
- remove shade from AWX requirements
- remove kombu from Ansible requirements

Other notes

Add note about pinning setuptools and pip,
done but not mentioned previously

Stop pinning gevent-websocket and twisted

upgrade Azure to Ansible core requirements

more detailed notes
https://gist.github.com/AlanCoding/9442a512ab6977940bc7b5b346d4f70b

upgrade version of Django for Exception

wenottingham commented:
Re: azure - just go with what's in https://github.com/ansible/ansible/blob/stable-2.5/packaging/requirements/requirements-azure.txt

AlanCoding commented:
Link #660

I had done 1.59 in the tower venv, but had missed the fact that I had to repeat this for the ansible one (clearly very important). After another cycle of this, I'll start trying to remove the hack.

Link #1144 (haven't looked much into yet)

@wenottingham That's great, it answers the case for the Ansible requirements. I'm not sure what the `azure` requirement in `requirements/requirements.in` is for though. Maybe I shouldn't touch it.

I'll get those tests fixed. But fist I need to fix a segmentation fault.

AlanCoding commented on requirements/requirements.in:
```diff
@@ -1,59 +1,59 @@
 djangorestframework-yaml==1.0.3
-gevent-websocket==0.9.5
-irc==15.1.1
+gevent-websocket==0.9.5   # pinned because of django channels
```
@chrismeyersfsu can I start putting comments to the side like this? This is the manually curated list, but I feel like anyone coming by would have no idea why anything was what it was.

matburt replied:
I don't think this is needed by django channels?  This was our old websocket interface requirement

https://github.com/django/channels/blob/master/setup.py#L14-L27

AlanCoding replied:
Let me ping @wwitzel3 about that.

But I'm happy to remove them and carry on. Some people have already asked me to verify that the UI websockets are working, so I'll do this anyway.

AlanCoding replied:
The comment in the README is

> Packages gevent-websocket and twisted are put in requirements.in not because they are primary dependency of Tower, but because their versions needs to be freezed as dependencies of django channel. Please be mindful when doing dependency updates.

I don't know if this still applies for current dependencies.

AlanCoding commented:
New one, don't know why that hadn't come up before.

```
awx_1        | Traceback (most recent call last):
awx_1        |   File "./awx/wsgi.py", line 33, in <module>
awx_1        |     application = get_wsgi_application()
awx_1        |   File "/venv/awx/lib/python2.7/site-packages/django/core/wsgi.py", line 13, in get_wsgi_application
awx_1        |     django.setup(set_prefix=False)
awx_1        |   File "/venv/awx/lib/python2.7/site-packages/django/__init__.py", line 27, in setup
awx_1        |     apps.populate(settings.INSTALLED_APPS)
awx_1        |   File "/venv/awx/lib/python2.7/site-packages/django/apps/registry.py", line 85, in populate
awx_1        |     app_config = AppConfig.create(entry)
awx_1        |   File "/venv/awx/lib/python2.7/site-packages/django/apps/config.py", line 94, in create
awx_1        |     module = import_module(entry)
awx_1        |   File "/usr/lib64/python2.7/importlib/__init__.py", line 37, in import_module
awx_1        |     __import__(name)
awx_1        |   File "/venv/awx/lib/python2.7/site-packages/social_django/__init__.py", line 4, in <module>
awx_1        |     from social_core.backends.base import BaseAuth
awx_1        | ImportError: No module named social_core.backends.base
```

(wait, this may be transient)

peterdemin commented:
Hey Alan, I guess you are running a lot of `pip-compile` in this task.
I want to suggest you trying [pip-compile-multi](https://github.com/peterdemin/pip-compile-multi) - it fixes a few pain points of AWX locking workflow. Also, I can add required changes if you don't mind.

AlanCoding commented:
@peterdemin Thanks for the tip. I would like to look into that, but right now I'm kind of trying to power through the main dependency update work because that can hold other people up. From the README there, it looks like there is a case that it more accurately fits what we're trying to do. I will try to experiment with it after the rest of this is ready.

AlanCoding commented:
Carded out the python-saml issue separately #1421

Adding a note on that REQUIREMENTS.in line for now, and moving on to resolve the other noted issues here.

AlanCoding commented:
Link https://github.com/robgolding/django-radius/issues/15

AlanCoding commented:
Most recent batch is in now.

This gets all of the tests running locally.

By manual testing, launching a job, websockets are still working @jakemcdermott 

Still need to finish going through the rest of the Ansible requirements to see if there's some requirements we can borrow from there.

AlanCoding commented:
Pushed a new batch.

Updated notes https://gist.github.com/AlanCoding/9442a512ab6977940bc7b5b346d4f70b

Updated Ansible requirements

Removed shade from one set, removed kombu from the other set.

Installing the new compiled dependencies now, will retest, run job after that finishes.

AlanCoding commented:
After latest batch & install, websockets are still working.

Shippable went green, so I think the boxes are all checked off now, removing WIP.

AlanCoding commented on requirements/requirements_isolated.txt:
```diff
@@ -1,3 +1,3 @@
-psutil==5.2.2
-pexpect==4.2.1
+psutil==5.4.3     # same as AWX requirement
+pexpect==4.4.0    # same as AWX requirement
```
@ryanpetrello looks like `pexpect` version seems to have diverged in the main requirements vs. the isolated requirements as-of the last set of dependency updates.

ryanpetrello replied:
d'oh!  This is likely my fault.  We want the latest of both - the newer pexpect includes my search optimizations from: https://github.com/pexpect/pexpect/pull/464

matburt commented:
Have we done some basic validation of inventory imports from gce, ec2, and azure?

AlanCoding commented:
Yep, good point. Re-running those got a new problem with boto now.

```
stderr:
ERROR! Could not parse inventory source /awx_devel/awx/plugins/inventory/ec2.py with available plugins:
Plugin script failed: Inventory script (/awx_devel/awx/plugins/inventory/ec2.py) had an execution error: Traceback (most recent call last):
  File "/awx_devel/awx/plugins/inventory/ec2.py", line 1701, in <module>
    Ec2Inventory()
  File "/awx_devel/awx/plugins/inventory/ec2.py", line 272, in __init__
    self.do_api_calls_update_cache()
  File "/awx_devel/awx/plugins/inventory/ec2.py", line 539, in do_api_calls_update_cache
    self.get_instances_by_region(region)
  File "/awx_devel/awx/plugins/inventory/ec2.py", line 605, in get_instances_by_region
    reservations = conn.get_all_instances()
  File "/venv/ansible/lib/python2.7/site-packages/boto/ec2/connection.py", line 585, in get_all_instances
    max_results=max_results)
  File "/venv/ansible/lib/python2.7/site-packages/boto/ec2/connection.py", line 681, in get_all_reservations
    [('item', Reservation)], verb='POST')
  File "/venv/ansible/lib/python2.7/site-packages/boto/connection.py", line 1170, in get_list
    response = self.make_request(action, params, path, verb)
  File "/venv/ansible/lib/python2.7/site-packages/boto/connection.py", line 1116, in make_request
    return self._mexe(http_request)
  File "/venv/ansible/lib/python2.7/site-packages/boto/connection.py", line 1030, in _mexe
    raise ex
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:579)

```

  boto v2.48.0  fails
  boto v2.47.0  works
  boto v2.46.1  works

AlanCoding commented:
Pinned the boto version... checked openstack, that's working without any other trouble.

AlanCoding commented on requirements/requirements.in:
```diff
@@ -1,59 +1,57 @@
+slackclient==1.1.2
 tacacs_plus==1.0
-twilio==6.1.0
-twisted==17.9.0
```
@wwitzel3 Halp! I don't know what I'm doing.

The websockets still work after I tested this. Does that mean we just don't need twisted and the gevent library anymore?

I'm pretty lost as to the necessity of these things on my own.

matburt replied:
I'd be surprised if we still needed gevent or twisted anymore.

wwitzel3 replied:
No, we do not need them

AlanCoding commented:
Okay, I just hit

```
RuntimeError: Django version other than 1.11.7 detected 1.11.10.             Inherit from WSGIHandler to support short-circuit Django Middelware.             This is known to work for Django 1.11.7 and may not work with other,             even minor, versions.
```

This was expected due to another PR, this just needs to be updated.

AlanCoding commented:
Pushed commit, since django-radius responded to my request to add future to their requirements I was able to remove it from our top-level requirements.

The pip compile commits are getting a little messed up, if I have to do any other major changes I'll blow away all these commits.

AlanCoding commented:
@shanemcd what are your thoughts on the mergability of this?

AlanCoding commented on requirements/requirements.in:
```diff
@@ -1,59 +1,56 @@
-apache-libcloud==2.0.0
+apache-libcloud==2.2.1
 appdirs==1.4.2
-asgi-amqp==1.1.1
```
Oh man, this was not intended. It must have been a wrong decision in a rebasing

AlanCoding commented:
Had a fairly major oversight where asgi-amqp was omitted from the requirements, which obviously shouldn't have been the case. That should be fixed now, commits squashed too.

peterdemin commented:
@AlanCoding is it a good time to introduce [pip-compile-multi](https://github.com/peterdemin/pip-compile-multi)?

shanemcd commented:
@peterdemin I've looked at your readme. What pain points do you see this alleviating exactly? If you are talking about the separate requirements files, they are not trying to create any sort of inheritance. They are used to create 2 separate virtual environments on the same system.

matburt commented:
From a strictly dependency tracking perspective I'd also rather head towards `pipenv`

peterdemin commented:
@shanemcd, I was referencing to pain points described in [Known Issues](https://github.com/ansible/awx/tree/devel/requirements#known-issues). I think all of the issues are solved, actually. It means less tedious handwork and oversights.
`pip-compile-multi` is perfectly suitable for a number of separate environments.
@matburt AFAIK `pipenv` doesn't support complex cases with multiple requirements files.
